common/lib/client/rest.ts

```diff
@@ -0,0 +1,225 @@
+import Platform from 'platform';
+import * as Utils from '../util/utils';
+import Logger, { LoggerOptions } from '../util/logger';
+import Defaults from '../util/defaults';
+import Auth from './auth';
+import Push from './push';
+import Http from 'platform-http';
+import PaginatedResource, { HttpPaginatedResponse, PaginatedResult } from './paginatedresource';
+import Channel from './channel';
+import ErrorInfo from '../types/errorinfo';
+import Stats from '../types/stats';
+import HttpMethods from '../../constants/HttpMethods';
+import { ChannelOptions } from '../../types/channel';
+import { PaginatedResultCallback, StandardCallback } from '../../types/utils';
+import { ErrnoException, RequestParams } from '../../types/http';
+import ClientOptions, { DeprecatedClientOptions, NormalisedClientOptions } from '../../types/ClientOptions';
+
+const noop = function() {};
+const msgpack = Platform.msgpack;
+
+class Rest {
+	options: NormalisedClientOptions;
+	baseUri: (host: string) => string;
+	authority: (host: string) => string;
+	_currentFallback: null | {
+		host: string,
+		validUntil: number,
+	};
+	serverTimeOffset: number | null;
+	auth: Auth;
+	channels: Channels;
+	push: Push;
+
+	constructor(options: ClientOptions | string) {
+		/* normalise options */
+		if(!options) {
+			const msg = 'no options provided';
+			Logger.logAction(Logger.LOG_ERROR, 'Rest()', msg);
+			throw new Error(msg);
+		}
+		const optionsObj = Defaults.objectifyOptions(options);
+
+		if(optionsObj.log) {
+			Logger.setLog(optionsObj.log.level, optionsObj.log.handler);
+		}
+		Logger.logAction(Logger.LOG_MICRO, 'Rest()', 'initialized with clientOptions ' + Utils.inspect(options));
+
+		const normalOptions = this.options = Defaults.normaliseOptions(optionsObj);
+
+		/* process options */
+		if(normalOptions.key) {
+			const keyMatch = normalOptions.key.match(/^([^:\s]+):([^:.\s]+)$/);
+			if(!keyMatch) {
+				const msg = 'invalid key parameter';
+				Logger.logAction(Logger.LOG_ERROR, 'Rest()', msg);
+				throw new Error(msg);
+			}
+			normalOptions.keyName = keyMatch[1];
+			normalOptions.keySecret = keyMatch[2];
```

Comment:
Access to array elements without check ☝️

Reply:
Inclined to agree but this is a pretty fundamental part of the library and hasn't caused a problem so far so I say we keep it for now

Reply:
Resolving this conversation on @ikbalkaya's behalf.